crates/aggchain-proof-service/src/service/customchaindata_builder.rs

```diff
@@ -0,0 +1,281 @@
+use std::{
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use futures::{future::BoxFuture, FutureExt as _};
+use tracing::error;
+
+use crate::context::AGGCHAIN_PROOF_PROGRAM_VERSION;
+
+const AGGCHAIN_TYPE: u16 = 0x0001;
+const AGGCHAIN_VKEY_SELECTOR: [u8; 4] =
+    calculate_aggchain_vkey_selector(AGGCHAIN_PROOF_PROGRAM_VERSION, AGGCHAIN_TYPE);
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {}
+
+#[derive(Debug)]
+pub(crate) struct CustomChainDataBuilderRequest {
+    pub(crate) l2_block_number: u64,
+    pub(crate) output_root: [u8; 32],
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) struct CustomChainDataBuilderResponse {
+    pub custom_chain_data: [u8; 66],
+}
+
+pub(crate) struct CustomChainDataBuilderService<L1Rpc> {
+    l1_rpc: Arc<L1Rpc>,
+    network_id: u32,
+    vkey: Arc<[u32; 8]>,
+}
+
+impl Clone for CustomChainDataBuilderService<L1Rpc> {
+    fn clone(&self) -> Self {
+        CustomChainDataBuilderService {
+            l1_rpc: self.l1_rpc.clone(),
+            network_id: self.network_id,
+            vkey: self.vkey.clone(),
+        }
+    }
+}
+
+impl<L1Rpc> CustomChainDataBuilderService<L1Rpc> {
+    pub fn new(l1_rpc: Arc<L1Rpc>, network_id: u32, vkey: [u32; 8]) -> Self {
+        CustomChainDataBuilderService {
+            l1_rpc,
+            network_id,
+            vkey: Arc::new(vkey),
+        }
+    }
+}
+
+impl<L1Rpc> tower::Service<CustomChainDataBuilderRequest> for CustomChainDataBuilderService<L1Rpc>
+where
+    L1Rpc: AggchainVkeyResolver + Send + Sync + 'static,
+{
+    type Response = CustomChainDataBuilderResponse;
+    type Error = Error;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, req: CustomChainDataBuilderRequest) -> Self::Future {
+        let rpc = self.l1_rpc.clone();
+
+        let vkey = self.vkey.clone();
```

Comment:
Seems vkey does not need to be cloned here. Taking a reference suffices for the comparison below.

Reply:
I don't think we can because of lifetime